File: src/main/java/voidchess/figures/Knight.java
package voidchess.figures;

import voidchess.board.BasicChessGameInterface;
import voidchess.board.SimpleChessBoardInterface;
import voidchess.helper.Move;
import voidchess.helper.Position;
import voidchess.image.ImageType;

import java.util.ArrayList;
import java.util.List;

/**
 * @author stephan
 */
public class Knight extends Figure {
    private List<Position> posIter = new ArrayList<Position>(8);

    public Knight(boolean isWhite, Position startPosition) {
        super(isWhite, startPosition, FigureType.KNIGHT);
    }

    public boolean isReachable(Position to, BasicChessGameInterface game) {
        int horizontal_difference = Math.abs(getPosition().getRow() - to.getRow());
        int vertical_difference = Math.abs(getPosition().getColumn() - to.getColumn());

        if (horizontal_difference + vertical_difference != 3
                || horizontal_difference * vertical_difference == 0) {
            return false;
        }

        return game.isFreeArea(to) || hasDifferentColor(game.getFigure(to));
    }

    private List<Position> getIterator(BasicChessGameInterface game) {
        posIter.clear();

        int minRow = Math.max(getPosition().getRow() - 2, 0);
        int minColumn = Math.max(getPosition().getColumn() - 2, 0);
        int maxRow = Math.min(getPosition().getRow() + 2, 7);
        int maxColumn = Math.min(getPosition().getColumn() + 2, 7);

        for (int row = minRow; row <= maxRow; row++) {
            for (int column = minColumn; column <= maxColumn; column++) {
                if (Math.abs(row - getPosition().getRow()) + Math.abs(column - getPosition().getColumn()) == 3) {
                    posIter.add(Position.Companion.get(row, column));
                }
            }
        }

        return posIter;
    }

    public void getReachableMoves(BasicChessGameInterface game, List<Move> result) {
        List<Position> posIter = getIterator(game);
        for (Position checkPosition : posIter) {
            if (isReachable(checkPosition, game)) {
                result.add(Move.Companion.get(getPosition(), checkPosition));
            }
        }
    }

    public boolean isSelectable(SimpleChessBoardInterface game) {
        List<Position> positions = getIterator(game);
        for (Position checkPosition : positions) {
            if (isMovable(checkPosition, game)) {
                return true;
            }
        }

        return false;
    }

    public int countReachableMoves(BasicChessGameInterface game) {
        return getIterator(game).size();
    }
}

File: src/main/kotlin/voidchess/figures/Knight.kt
package voidchess.figures

import voidchess.board.BasicChessBoard
import voidchess.board.ChessBoard
import voidchess.board.check.BoundLine
import voidchess.board.check.CheckLine
import voidchess.board.move.Move
import voidchess.board.move.Position


class Knight(isWhite: Boolean, startPosition: Position) : Figure(isWhite, startPosition, FigureType.KNIGHT, false, false) {

    override fun isReachable(toPos: Position, game: BasicChessBoard): Boolean {
        val horizontalDifference = Math.abs(position.row - toPos.row)
        val verticalDifference = Math.abs(position.column - toPos.column)

        if (horizontalDifference + verticalDifference != 3 || horizontalDifference == 0 || verticalDifference == 0) {
            return false
        }

        val figure = game.getFigureOrNull(toPos)
        return figure == null || hasDifferentColor(figure)
    }

    override fun getReachableMoves(game: BasicChessBoard, result: MutableList<Move>) {
        forEachReachablePos(game) {
            result.add(Move[position, it])
        }
    }

    override fun getPossibleMovesWhileUnboundAndCheck(game: ChessBoard, checkLine: CheckLine, result: MutableList<Move>) {
        // the accessibility of the target field doesn't need to be checked because
        // all checkInterceptPositions are guaranteed to be either empty
        // or to contain the attacker (who has a different color)
        checkLine.posProgression.forEachReachablePos {checkInterceptPos->
            val horizontalDifference = Math.abs(position.row - checkInterceptPos.row)
            val verticalDifference = Math.abs(position.column - checkInterceptPos.column)
            if(horizontalDifference+verticalDifference==3 && horizontalDifference!=0 && verticalDifference!=0) {
                result.add(Move[position, checkInterceptPos])
            }
        }
    }

    override fun getPossibleMovesWhileBoundAndNoCheck(game: ChessBoard, boundLine: BoundLine, result: MutableList<Move>) {
        // a bound move can't move at all!
    }

    override fun isSelectable(game: ChessBoard): Boolean {
        forEachReachablePos(game) {
            if (!isBound(it, game)) return true
        }
        return false
    }

    override fun countReachableMoves(game: BasicChessBoard): Int {
        var reachableMovesCount = 0
        forEachReachablePos(game) {
            reachableMovesCount++
        }
        return reachableMovesCount
    }

    private inline fun forEachReachablePos(game: BasicChessBoard, informOf: (Position) -> Unit) {
        position.forEachKnightPos { pos ->
            if(isAccessible(game, pos)) informOf(pos)
        }
    }
}
